godot-macros/src/class/godot_api.rs

```diff
@@ -408,49 +409,76 @@ fn transform_trait_impl(original_impl: Impl) -> Result<TokenStream, Error> {
             continue;
         };
 
+        // Transport #[cfg] attributes to the virtual method's FFI glue, to ensure it won't be
+        // registered in Godot if conditionally removed from compilation.
+        let cfg_attrs = util::extract_cfg_attrs(&method.attributes)
+            .into_iter()
+            .collect::<Vec<_>>();
         let method_name = method.name.to_string();
         match method_name.as_str() {
             "register_class" => {
                 register_class_impl = quote! {
+                    #(#cfg_attrs)*
                     impl ::godot::obj::cap::GodotRegisterClass for #class_name {
                         fn __godot_register_class(builder: &mut ::godot::builder::GodotBuilder<Self>) {
                             <Self as #trait_name>::register_class(builder)
                         }
                     }
                 };
 
+                // Use 'match' as a way to only return 'Some(...)' if the given cfg attrs allow.
+                // Needs '#[allow(unreachable_patterns)]' to avoid warnings about the last arm.
                 register_fn = quote! {
-                    Some(#prv::ErasedRegisterFn {
-                        raw: #prv::callbacks::register_class_by_builder::<#class_name>
-                    })
+                    match () {
+                        #(#cfg_attrs)*
+                        () => Some(#prv::ErasedRegisterFn {
+                            raw: #prv::callbacks::register_class_by_builder::<#class_name>
+                        }),
+                        _ => None,
+                    }
```

Comment:
This is a very cool trick! 

Do you not have to add `#[allow(unreachable_patterns)]` already here? The more local, the better.

Also, does clippy not complain if the `#[cfg]` evaluates to false and the remaining code becomes
```rs
match () {
   _ => None,
}
```
Or maybe even without that, having a `match ()` which can only have one value?

Reply:
> This is a very cool trick!
Thanks!

> Can you add `#[allow(unreachable_patterns)]` right here? The more local, the better -- otherwise we may accidentally swallow legitimate warnings in unrelated code.

Unfortunately, since the `match` is being used as an expression, we can't due to https://github.com/rust-lang/rust/issues/15701 (currently results in an error). Adding macros above match arms does compile, but for this purpose is useless as `#[allow(...)]` over a single match arm doesn't dismiss any related warnings.

> Also, does clippy not complain if the `#[cfg]` evaluates to false and the remaining code becomes
> 
> ```rust
> match () {
>    _ => None,
> }
> ```

Great catch:

```
warning: this match could be replaced by its scrutinee and body
   --> examples/dodge-the-creeps/rust/src/main_scene.rs:103:31
    |
103 |           let _x: Option<i32> = match () {
    |  _______________________________^
104 | |             _ => None,
105 | |         };
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_single_binding
    = note: `#[warn(clippy::match_single_binding)]` on by default
help: consider using the scrutinee and body instead
    |
103 ~         let _x: Option<i32> = ();
104 ~         None;
    |
```

I've added `#[allow(clippy::match_single_binding)]` as well now.

> Or maybe even without that, having a `match ()` which can only have one value?

That doesn't seem to generate a warning by itself.

Reply:
Thanks a lot! You now have this:
```rs
            #[allow(unreachable_patterns)] // due to the cfg-based match statements
            #[allow(clippy::match_single_binding)] // avoid warning on single-arm matches
```
Could you clarify the 2 comments slightly? e.g. the 2nd could be "for #[cfg(false)], match contains only single arm".

Reply:
Okay, I've tried to improve the comments a bit; feel free to suggest further changes.

Also, I managed to move the `allow`s to the `match` itself by placing the match inside a block, as showcased below:

```rs
fn main() {
    let x = {
        #[allow(unreachable_patterns)]
        #[allow(clippy::match_single_binding)]
        match () {
            () => true,
            _ => false,
        }
    };
    // ^instead of let x = match () { ... }
    println!("x = {x}");
}
```

I'm not sure why this works - seems to work even with something as simple as just `5i32`, so perhaps it's some form of workaround for the aforementioned Rust compiler issue. The sample above (on my limited local tests) seems to compile with Rust 1.69.0 as well, so this shouldn't raise MSRV.